src/timepicker/TimePicker.ts

```diff
@@ -0,0 +1,355 @@
+
+			if (label) {
+				children = [ w(Label, {
+					classes: this.classes(
+						disabled ? css.disabled : null,
+						invalid ? css.invalid : null,
+						readOnly ? css.readonly : null,
+						required ? css.required : null
+					),
+					formId,
+					label
+				}, [ input ]) ];
+			}
+
+			return v('span', {
+				classes: this.classes(css.root),
+				key: 'root'
+			}, children);
+		}
+
+		return this.renderCustomInput();
+	}
+
+	@onPropertiesChanged()
+	protected onPropertiesChanged(event: PropertiesChangeEvent<this, TimePickerProperties>) {
+		if (
+			!this._getOptions ||
+			includes(event.changedPropertyKeys, 'start') ||
+			includes(event.changedPropertyKeys, 'end') ||
+			includes(event.changedPropertyKeys, 'step')) {
+			this._options = null;
+		}
+	}
+
+	protected renderNativeInput() {
+		const {
+			disabled,
+			end,
+			invalid,
+			name,
+			readOnly,
+			required,
+			start,
+			step,
+			value
+		} = this.properties;
+
+		const classes = [
+			css.input,
+			disabled ? css.disabled : null,
+			invalid ? css.required : null,
+			readOnly ? css.readonly : null,
+			required ? css.required : null
+		];
+
+		return v('input', {
+			'aria-invalid': invalid ? 'true' : null,
+			'aria-readonly': readOnly ? 'true' : null,
+			classes: this.classes(...classes),
+			disabled,
+			invalid,
+			key: 'native-input',
+			max: end,
+			min: start,
+			name,
+			onblur: this._onNativeBlur,
+			onchange: this._onNativeChange,
+			onfocus: this._onNativeFocus,
+			readOnly,
+			required,
+			step,
+			type: 'time',
+			value
+		});
+	}
+
+	protected renderCustomInput() {
+		const {
+			autoBlur,
+			clearable,
+			customOptionItem,
+			customOptionMenu,
+			disabled,
+			extraClasses,
+			formId,
+			inputProperties,
+			invalid,
+			isOptionDisabled,
+			label,
+			onBlur,
+			onChange,
+			onFocus,
+			onMenuChange,
+			openOnFocus,
+			options,
+			readOnly,
+			required,
+			theme,
+			value
+		} = this.properties;
+
+		return w(ComboBox, {
+			autoBlur,
+			clearable,
+			customResultItem: customOptionItem,
+			customResultMenu: customOptionMenu,
+			disabled,
+			extraClasses,
+			formId,
+			getResultLabel: this._getOptionLabel,
+			inputProperties,
+			invalid,
+			isResultDisabled: isOptionDisabled,
+			label,
+			onBlur,
+			onChange,
+			onFocus,
+			onMenuChange,
+			onRequestResults: this._onRequestOptions,
+			openOnFocus,
+			readOnly,
+			required,
+			results: options,
+			theme,
+			value
+		});
+	}
+
+	private _formatUnits(units: TimeUnits): string {
+		const { step = 60 } = this.properties;
+		const { hour, minute, second } = units;
+
+		return (step >= 60 ? [ hour, minute ] : [ hour, minute, second ])
+			.map(unit => padStart(String(unit), 2, '0'))
+			.join(':');
+	}
+
+	private _getOptionLabel(value: TimeUnits) {
+		const { getOptionLabel } = this.properties;
+		const units = parseUnits(value);
+		return getOptionLabel ? getOptionLabel(units) : this._formatUnits(units);
+	}
+
+	private _getOptions() {
```

Comment:
Hm, I see the benefit of caching the results, but I'm not sure it's necessary to avoid passing back the actual options array. If someone wants to use a different method to generate options, I think they should just extend the widget. @bitpshr do you have an opinion one way or the other?